File: apps/api/routers/inhouse.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, status

from core.deps import require_librarian
from core.supabase import get_admin_client
from schemas.auth import UserProfile
from schemas.inhouse import CreateInHouseLoanRequest, InHouseLoan

router = APIRouter(prefix="/in-house-loans", tags=["in-house-loans"])

# Plan 7: entirely librarian-driven — a guest has no session, no JWT, and
# never calls any endpoint in this router themselves.

@router.get("", response_model=list[InHouseLoan])
def list_in_house_loans(
    status_filter: str | None = None,
    librarian: UserProfile = Depends(require_librarian),
):
    db = get_admin_client()
    query = db.table("in_house_loans").select("*, book_copies(accession_number, books(*))").order("checked_out_at", desc=True)
    if status_filter:
        query = query.eq("status", status_filter)
    res = query.execute()
    loans = res.data
    for loan in loans:
        copy = loan.pop("book_copies", None)
        loan["books"] = copy["books"] if copy else None
        loan["accession_number"] = copy["accession_number"] if copy else None
    return loans

@router.post("", response_model=InHouseLoan, status_code=status.HTTP_201_CREATED)
def create_in_house_loan(
    body: CreateInHouseLoanRequest,
    librarian: UserProfile = Depends(require_librarian),
):
    db = get_admin_client()

    copy_res = db.table("book_copies").select("*").eq("accession_number", body.accession_number.strip()).execute()
    if not copy_res.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "No copy with that accession number")
    copy = copy_res.data[0]
    if copy["status"] != "available":
        raise HTTPException(status.HTTP_400_BAD_REQUEST, f"This copy is currently '{copy['status']}', not available")

    if body.visitor_type == "non_nocei" and not body.fee_paid:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "The ₱50.00 visitor fee must be collected before checkout")

    loan_res = db.table("in_house_loans").insert({
        "book_copy_id": copy["id"],
        "librarian_id": librarian.id,
        "guest_name": body.guest_name.strip(),
        "guest_id_number": body.guest_id_number.strip(),
        "visitor_type": body.visitor_type,
        "fee_paid": body.fee_paid,
        "purpose": body.purpose,
        "notes": body.notes,
    }).execute()
    if not loan_res.data:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Could not create the in-house loan")

    # available -> on_loan: legal per Phase 1's status machine. Correctly
    # takes the copy out of circulation while the guest has it, exactly
    # like a real loan does — it's just not going home with them.
    db.table("book_copies").update({"status": "on_loan"}).eq("id", copy["id"]).execute()

    loan = loan_res.data[0]
    book_res = db.table("books").select("*").eq("id", copy["book_id"]).execute()
    loan["books"] = book_res.data[0] if book_res.data else None
    loan["accession_number"] = copy["accession_number"]
    return loan

@router.post("/{loan_id}/return", response_model=InHouseLoan)
def return_in_house_loan(
    loan_id: str,
    librarian: UserProfile = Depends(require_librarian),
):
    db = get_admin_client()

    loan_res = db.table("in_house_loans").select("*").eq("id", loan_id).execute()
    if not loan_res.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "In-house loan not found")
    loan = loan_res.data[0]
    if loan["status"] == "returned":
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "This item has already been returned")

    update_res = db.table("in_house_loans").update({
        "status": "returned",
        "returned_at": datetime.now(timezone.utc).isoformat(),
    }).eq("id", loan_id).execute()
    loan = update_res.data[0]

    # on_loan -> for_reshelving: never straight back to available, same
    # rule Phase 4 follows for real returns (build plan 4.6/4.7).
    copy_res = db.table("book_copies").select("id, book_id, accession_number").eq("id", loan["book_copy_id"]).execute()
    copy = copy_res.data[0]
    db.table("book_copies").update({"status": "for_reshelving"}).eq("id", copy["id"]).execute()

    book_res = db.table("books").select("*").eq("id", copy["book_id"]).execute()
    loan["books"] = book_res.data[0] if book_res.data else None
    loan["accession_number"] = copy["accession_number"]
    return loan

File: apps/web/lib/inhouse.ts
// apps/web/lib/inhouse.ts
// Fetch layer for the librarian-side guest / in-house loan flow (kiosk
// plan Phase 7). Entirely librarian-driven — a guest never calls any of
// this themselves, they have no session and no JWT.

import { getToken } from "@/lib/auth"
import type { Book } from "@lasallia/types"

const API_URL = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"

async function parseErrorOrThrow(res: Response, fallback: string): Promise<never> {
  const body = await res.json().catch(() => ({}))
  throw new Error(body.detail ?? fallback)
}

function authHeaders(): HeadersInit {
  const token = getToken()
  if (!token) throw new Error("Not signed in")
  return { Authorization: `Bearer ${token}` }
}

export type VisitorType = "nocei" | "non_nocei"
export type Purpose = "library_use" | "photocopy"

export type InHouseLoan = {
  id: string
  book_copy_id: string
  accession_number: string | null
  librarian_id: string
  guest_name: string
  guest_id_number: string
  visitor_type: VisitorType
  fee_paid: boolean
  purpose: Purpose
  checked_out_at: string
  returned_at: string | null
  status: "active" | "returned"
  notes: string | null
  books: Book | null
}

export async function fetchInHouseLoans(status?: "active" | "returned"): Promise<InHouseLoan[]> {
  const qs = status ? `?status_filter=${status}` : ""
  const res = await fetch(`${API_URL}/in-house-loans${qs}`, { headers: authHeaders() })
  if (!res.ok) return parseErrorOrThrow(res, "Failed to load in-house loans")
  return res.json()
}

export async function createInHouseLoan(params: {
  accessionNumber: string
  guestName: string
  guestIdNumber: string
  visitorType: VisitorType
  feePaid: boolean
  purpose: Purpose
  notes?: string
}): Promise<InHouseLoan> {
  const res = await fetch(`${API_URL}/in-house-loans`, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify({
      accession_number: params.accessionNumber,
      guest_name: params.guestName,
      guest_id_number: params.guestIdNumber,
      visitor_type: params.visitorType,
      fee_paid: params.feePaid,
      purpose: params.purpose,
      notes: params.notes || undefined,
    }),
  })
  if (!res.ok) return parseErrorOrThrow(res, "Could not check out this item")
  return res.json()
}

export async function returnInHouseLoan(id: string): Promise<InHouseLoan> {
  const res = await fetch(`${API_URL}/in-house-loans/${id}/return`, {
    method: "POST",
    headers: authHeaders(),
  })
  if (!res.ok) return parseErrorOrThrow(res, "Could not confirm this return")
  return res.json()
}
